src/main/java/com/thealgorithms/sorts/CountingSort.java

```diff
@@ -1,93 +1,53 @@
     @Override
     public <T extends Comparable<T>> List<T> sort(List<T> list) {
-        Map<T, Integer> frequency = new TreeMap<>();
-        // The final output array
-        List<T> sortedArray = new ArrayList<>(list.size());
-
-        // Counting the frequency of @param array elements
-        list.forEach(v -> frequency.put(v, frequency.getOrDefault(v, 0) + 1));
+        Map<T, Integer> frequencyMap = computeFequencyMap(list);
+        return extractSortedArray(frequencyMap);
+    }
 
-        // Filling the sortedArray
-        for (Map.Entry<T, Integer> element : frequency.entrySet()) {
-            for (int j = 0; j < element.getValue(); j++) {
-                sortedArray.add(element.getKey());
+    private static <T extends Comparable<T>> List<T> extractSortedArray(Map<T, Integer> frequencyMap) {
+        List<T> sortedList = new ArrayList<>();
+        for (Map.Entry<T, Integer> entry : frequencyMap.entrySet()) {
+            for (int i = 0; i < entry.getValue(); i++) {
+                sortedList.add(entry.getKey());
             }
         }
-
-        return sortedArray;
-    }
-
-    /**
-     * Stream Counting Sort The same as method {@link CountingSort#sort(List)} }
-     * but this method uses stream API
-     *
-     * @param list The list to be sorted
-     */
-    private static <T extends Comparable<T>> List<T> streamSort(List<T> list) {
-        return list.stream().collect(toMap(k -> k, v -> 1, (v1, v2) -> v1 + v2, TreeMap::new)).entrySet().stream().flatMap(entry -> IntStream.rangeClosed(1, entry.getValue()).mapToObj(t -> entry.getKey())).collect(toList());
+        return sortedList;
     }
 
-    // Driver Program
-    public static void main(String[] args) {
-        // Integer Input
-        List<Integer> unsortedInts = Stream.of(4, 23, 6, 78, 1, 54, 23, 1, 9, 231, 9, 12).collect(toList());
-        CountingSort countingSort = new CountingSort();
-
-        System.out.println("Before Sorting:");
-        print(unsortedInts);
-
-        // Output => 1 1 4 6 9 9 12 23 23 54 78 231
-        System.out.println("After Sorting:");
-        print(countingSort.sort(unsortedInts));
-        System.out.println("After Sorting By Streams:");
-        print(streamSort(unsortedInts));
-
-        System.out.println("\n------------------------------\n");
-
-        // String Input
-        List<String> unsortedStrings = Stream.of("c", "a", "e", "b", "d", "a", "f", "g", "c").collect(toList());
-
-        System.out.println("Before Sorting:");
-        print(unsortedStrings);
-
-        // Output => a a b c c d e f g
-        System.out.println("After Sorting:");
-        print(countingSort.sort(unsortedStrings));
-
-        System.out.println("After Sorting By Streams:");
-        print(streamSort(unsortedStrings));
+    private static <T extends Comparable<T>> Map<T, Integer> computeFequencyMap(List<T> list) {
+        // TreeMap to maintain order of elements naturally.
+        Map<T, Integer> frequencyMap = new TreeMap<>();
```

Comment:
I think that in the context of _cout sort_ it seams more natural to have a hash map (amortised $O(n)$ vs. $O(n \log n)$). 

```suggestion
        Map<T, Integer> frequencyMap = new HashMap<>();
```

Reply:
This TreeMapprovide us with ordering of keys, formation of a sorted array. And also helps to avoid standard code for countSort:

```
    public static int[] sort(int[] array) {
        if (array.length == 0) {
            return array;
        }

        int max = array[0];
        int min = array[0];

        for (int value : array) {
            if (value > max) {
                max = value;
            }
            if (value < min) {
                min = value;
            }
        }

        int range = max - min + 1;
        int[] count = new int[range];

        for (int value : array) {
            count[value - min]++;
        }

        int[] output = new int[array.length];
        int index = 0;

        for (int i = 0; i < range; i++) {
            while (count[i] > 0) {
                output[index++] = i + min;
                count[i]--;
            }
        }

        return output;
    }
```

So if it is removed, we will have to write another sorting implementation.

Reply:
This class does not implement [counting sort](https://en.wikipedia.org/wiki/Counting_sort).

Reply:
Yes, but this class already was before my changes, in sorting algorithms, and I just wanted to make some refactoring, to make it more corresponding according to other sorting algorithms, to general approach. If we will remove TreeMap this current approach will break. Possible options can be rename this class to CountingSortUsingTreeMap and create separate traditional implementation of it.

What you think, we need todo?

Reply:
Your update is a absolutely a good thing to do.

The implementation from the very beginning was not a _counting sort_.

One thing to do is to simply... delete this class. The reasons for that is simply, that in certain situations, the implementation can lead to strange bugs. Imagine that you work with a type `IntWithId`, which has two fields: `value` and `id`. `value` is used to compare the objects of that class. Consider the input like `[(4, 1), (3, 2), (3, 3), (3, 4)]`. The output will be like [(3, 2), (3, 2), (3, 2), (4, 1)].

Second argument: imagine someone preparing for an exam about sorting algorithms. And then, on the exam, there is a question _What is counting sort?_

Reply:
Educational exam case fully describe complexity of naming situation.

I don't know resolve my last changes the issue. But I have added canonical counting sort, which is pretty standard for any educational purposes. 

And renamed previous sorting with index Pseudo to highlight that it uses counting sort idea but it doesn't match with counting sort.